documentation/projects/proposals/trust_and_safety/content_report_moderation/20240122-implementation_plan_bulk_moderation_actions.md

```diff
@@ -0,0 +1,531 @@
+
+- [Project Thread](https://github.com/WordPress/openverse/issues/383)
+- [Project Proposal](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20230411-project_proposal_content_report_moderation.html#bulk-moderation-actions-requirement-3)
+
+## Overview
+
+<!-- An overview of the implementation plan, if necessary. Save any specific steps for the section(s) below. -->
+
+Typically, moderation actions are applied to individual records in response to a
+Report (see this
+[IP](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.html)
+for more details). However it may occasionally be necessary to perform a
+moderation action to a large set of records at once.
+
+For example, after receiving a large number of user reports for images from a
+particular creator, it may be beneficial to quickly mark all results from that
+creator as sensitive, or else deindex them entirely. Because these bulk actions
+necessarily affect large numbers of records without careful individual review,
+it's also necessary to be able to _undo_ these actions (for example, if further
+investigation reveals that only a subset of the records needs to be marked
+sensitive). We should also be able to view these bulk decisions as part of the
+complete moderation history of an individual record.
+
+This IP extends the work described in the
+[Django admin moderator access control and base improvements](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.html)
+IP to add support for these features. I have attempted to summarize all relevant
+parts of that IP as needed here, but for full context that IP should be
+considered prerequisite reading. For brevity it is referred to as "the related
+IP" hereafter.
+
+### Models
+
+No new models need be added; rather, we will make some small changes to the
+`ModerationDecision` model described in
+[this section](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20231208-implementation_plan_django_admin_moderator_access.html#moderationdecision)
+of the related IP. See the [Alternatives section](#alternatives) of this
+document for a discussion of why we choose this route, rather than adding new
+models.
+
+As a reminder, these are the fields from the `ModerationDecision` model which
+will not be changed:
+
+- `id`
+- `created_on`
+- `moderator_id`
+- `explanation`
+
+The following changes will be made to the `media_identifier` and `action`
+fields:
+
+#### `media_identifiers`
+
+`media_identifier` (previously a one-to-one relationship to the media the
+decision is related to), will be updated to `media_identifiers` _plural_: a
+many-to-many relationship to potentially multiple media records the decision is
+related to. Under the covers, this will be implemented by a join table. As
```

Comment:
Ah yes, that makes sense. I had not thought about the impact this would have on querying. But wouldn't DB-writes be incredibly slow if we have to write individual records for each media item in a bulk-moderation event?

Reply:
I don't believe so -- we can use [RelatedManager.set](https://docs.djangoproject.com/en/5.0/ref/models/relations/#django.db.models.fields.related.RelatedManager.set) to bulk set the list of related records, if I'm understanding your question!